File: typescript/src/Backtracking/Permutations/Permutations.ts
class Permutations {
  nums: number[]

  constructor(nums: number[]) {
    this.nums = nums
  }

  exec(): number[][] {
    return this.permute(this.nums)
  }

  exec2(): number[][]{
    return this.permute2(this.nums)
  }

  permute2(nums: number[]): number[][]{
    // special case handle
    if(nums.length === 1) return [[...nums]]

    const res: number[][] = []
    const LEN: number = nums.length;

    for(let i = 0; i < LEN; i++){
      const numsCopy: number[] = [...nums]
      // remove ith element
      numsCopy.splice(i, 1)
      // get permutations of the new array, and add the removed element back
      this.permute2(numsCopy).map(permutation => res.push([...permutation, nums[i]]))
    }

    return res;
  }

  permute(nums: number[]): number[][] {
    let res: number[][] = [];

    if(nums.length === 1) {
      return [[...nums]]
    }

    // for an array of length n, loop n times.
    // eslint-disable-next-line no-unused-vars
    for(let _ of nums){
      // remove the first value
      // @ts-ignore
      const first: number = nums.shift()
      // get permutation of remaining array
      const perms: number[][] = this.permute(nums)
      // add first value back
      for(let perm of perms){
        perm.push(first)
      }
      // update result
      res.push(...perms)
      // move first value to the back
      nums.push(first)
    }
    return res;
  }
}

export default Permutations

File: python/src/backtracking/permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # base case
        if len(nums) == 1:
            return [nums.copy()]

        # hold the final result
        res = []

        # we iterate through each number in list and remove it
        for i in range(len(nums)):
            nums_copy = nums.copy()
            nums_copy.pop(i)
            # get the remaining permutation and append the removed value in the end
            for temp in self.permute(nums_copy):
                temp.append(nums[i])
                res.append(temp)

        return res


solution = Solution()
print(solution.permute([1, 2]))
